rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeTypedArrayView.java

```diff
@@ -407,59 +455,94 @@ protected static NativeTypedArrayView<?> js_constructor(
+        int srcLength = source.length;
+
+        if (dbloff > targetLength) {
+            throw ScriptRuntime.rangeErrorById("msg.typed.array.bad.offset", dbloff);
+        }
+
+        if (srcLength + dbloff > targetLength) {
+            throw ScriptRuntime.rangeErrorById("msg.typed.array.bad.source.array");
+        }
+
+        if (isBigInt != source.isBigInt) {
+            throw ScriptRuntime.typeErrorById("msg.typed.array.type.mismatch");
         }
 
-        if (v.arrayBuffer == arrayBuffer) {
+        int targetOffset = (int) dbloff;
+        if (source.arrayBuffer == arrayBuffer) {
             // Copy to temporary space first, as per spec, to avoid messing up overlapping copies
-            Object[] tmp = new Object[v.length];
-            for (int i = 0; i < v.length; i++) {
-                tmp[i] = v.js_get(i);
+            Object[] tmp = new Object[srcLength];
+            for (int i = 0; i < srcLength; i++) {
+                tmp[i] = source.js_get(i);
             }
-            for (int i = 0; i < v.length; i++) {
-                js_set(i + off, tmp[i]);
+            for (int i = 0; i < srcLength; i++) {
+                js_set(i + targetOffset, tmp[i]);
             }
         } else {
-            for (int i = 0; i < v.length; i++) {
-                js_set(i + off, v.js_get(i));
+            for (int i = 0; i < srcLength; i++) {
+                js_set(i + targetOffset, source.js_get(i));
             }
         }
     }
 
-    private void setRange(NativeArray a, double dbloff) {
-        if (dbloff < 0 || dbloff > length) {
-            String msg = ScriptRuntime.getMessageById("msg.typed.array.bad.offset", dbloff);
-            throw ScriptRuntime.rangeError(msg);
+    private void setRange(Context cx, Scriptable scope, Scriptable source, double dbloff) {
+        if (isTypedArrayOutOfBounds()) {
+            throw ScriptRuntime.typeErrorById("msg.typed.array.out.of.bounds");
         }
-        int off = (int) dbloff;
-        if ((off + a.size()) > length) {
-            String msg = ScriptRuntime.getMessageById("msg.typed.array.bad.source.array");
-            throw ScriptRuntime.rangeError(msg);
+        int targetLength = length;
+        Scriptable src = ScriptRuntime.toObject(scope, source);
+        long srcLength = AbstractEcmaObjectOperations.lengthOfArrayLike(cx, src);
+
+        if (dbloff > targetLength) {
+            throw ScriptRuntime.rangeErrorById("msg.typed.array.bad.offset", dbloff);
         }
 
-        int pos = off;
-        for (Object val : a) {
-            js_set(pos, val);
-            pos++;
+        if (srcLength + dbloff > targetLength) {
+            throw ScriptRuntime.rangeErrorById("msg.typed.array.bad.source.array");
+        }
+
+        int targetOffset = (int) dbloff;
+
+        for (int k = 0; k < srcLength; k++) {
+            Object value = source.get(k, source);
+            js_set(k + targetOffset, value);
         }
     }
 
+    public boolean isTypedArrayOutOfBounds() {
```

Comment:
This feels like it's going to be an awfully hot code path, so I think it's worth thinking through this bit carefully. Since we do not yet support growable array buffers (and I'd expect growing a buffer to be much rarer than these checks) it might be worth calculating all this stuff when we create the typed array view, putting it in a boolean, and then just returning `arrayBuffer.isDetached() || outOfRange`. Okay, you'd want to make `arrayBufffer`, `offset`, and `byteLength` final, but that's a very small refactor.